### DISCUSSION TITLE
Check permission

### DISCUSSION THREAD

**Discussion by ISulimanov** [casdoor/casdoor#1681] [Q&A] [answered]
Lets say a have a role - devops and Permission - deploy with Sub roles devops.
User has role devops, but access token contains information only about role, and there is no info to figure out Is there access to deploy just from token.

What I right way to do it ?

**Comment by hsluoyz** [comment, marked as answer]
@ISulimanov Casdoor uses Casbin for advanced authorization: https://casbin.org/ , this is the general background

For your specific question, you have two ways to get permissions for a user:

1. Use the permissions inside the access token, so you can "list" the permissions the user has

https://github.com/casdoor/casdoor/blob/337ee2faefb8c80f9eb065696123e1ef6da02946/object/token_jwt.go#L172

and docs: https://casdoor.org/docs/permission/permission-configuration

2. Call Casbin Enforce API (a RESTful API) exposed by Casdoor. Casdoor will use Casbin engine to "check" the specific permission in a more accurate and "Casbin" way: https://casdoor.org/docs/permission/exposed-casbin-apis

You can choose either solution based on your scenario (rely on more Casbin or less Casbin)
